lib/rom/plugins/relation/sql/auto_restrictions.rb

```diff
@@ -0,0 +1,66 @@
+module ROM
+  module Plugins
+    module Relation
+      module SQL
+        # Generates methods for restricting relations by their indexed attributes
+        #
+        # This plugin must be enabled for the whole adapter, `use` won't work as
+        # schema is not yet available, unless it was defined explicitly.
+        #
+        # @example
+        #   rom = ROM.container(:sql, 'sqlite::memory') do |config|
+        #     config.create_table(:users) do
+        #       primary_key :id
+        #       column :email, String, null: false, unique: true
+        #     end
+        #
+        #     config.plugin(:sql, relations: :auto_restrictions)
+        #
+        #     config.relation(:users) do
+        #       schema(infer: true)
+        #     end
+        #   end
+        #
+        #   # now `by_email` is available automatically
+        #   rom.relations[:users].by_email('jane@doe.org')
+        #
+        # @api public
+        module AutoRestrictions
+          def self.included(klass)
+            super
+            methods, mod = restriction_methods(klass)
+            klass.include(mod)
+            methods.each { |meth| klass.auto_curry(meth) }
+          end
+
+          def self.restriction_methods(klass)
+            mod = Module.new
+
+            indexed_attrs = klass.schema.select { |attr| attr.meta[:index] }
+
+            methods = indexed_attrs.map do |attr|
+              attr_name = attr.name
+              meth_name = :"by_#{attr_name}"
+
+              mod.module_eval do
+                define_method(meth_name) do |value|
+                  where(self.class.schema[attr_name].is(value))
```

Comment:
Oh very good point, it can totally use the attr we already have. I'll change this, thanks